### PR TITLE
Comments on nodes

### PR DESCRIPTION
**Pull request by mwu-tow**
### Checklist
Please include the following checklist in your PR:

- [x] The `CHANGELOG.md` was updated with the changes introduced in this PR.
- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guide.
- [ ] All code has automatic tests where possible.
- [ ] All code has been profiled where possible.
- [x] All code has been manually tested in the IDE.
- [x] All code has been manually tested in the "debug/interface" scene.
- [ ] All code has been manually tested by the PR owner against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).
- [ ] All code has been manually tested by at least one reviewer against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).

### CODE REVIEW

File: src/rust/ide/lib/ast/impl/src/macros.rs [hunk 1]
```diff
@@ -8,21 +8,132 @@ use crate::crumbs::AmbiguousCrumb;
...
+
+// ================
+// === Comments ===
+// ================
+
+// === Disable Comments ===
+
+/// Try Interpreting the line as disabling comment. Return the text after `#`.
+pub fn as_disable_comment(ast:&Ast) -> Option<String> {
+    let r#match = crate::known::Match::try_from(ast).ok()?;
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/macros.rs]: Try without r + alignment

File: src/rust/ide/lib/ast/impl/src/macros.rs [hunk 2]
```diff
@@ -8,21 +8,132 @@ use crate::crumbs::AmbiguousCrumb;
...
+        // This gets us documentation text, however non-first lines have indent whitespace
+        // maintained.
+        let repr   = self.body.repr();
+        let indent = self.block_indent + DOCUMENTATION_COMMENT_INTRODUCER.len();
+        let old    = format!("\n{}", " ".repeat(indent));
+        let new    = "\n";
+        repr.replace(&old,new)
+    }
+
+    /// Get the documentation text.
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/macros.rs]: Docs are the same, and pretty is not a good word here.

File: src/rust/ide/src/double_representation/comment.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+use crate::prelude::*;
+
+// FIXME move to other parser-related tests
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation/comment.rs]: Remember to remove ALL Fixmes

File: src/rust/ide/src/double_representation/comment.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ast::Shape;
+    use crate::double_representation::definition::DefinitionProvider;
+    use ast::macros::DocCommentInfo;
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation/comment.rs]: Fix imports

File: src/rust/ide/src/double_representation.rs [hunk 5]
```diff
@@ -1,7 +1,16 @@
...
 //! module.
 
+use crate::prelude::*;
+
+use ast::{Ast, opr, prefix, known};
+use crate::double_representation::definition::DefinitionName;
+use crate::double_representation::definition::ScopeKind;
+use ast::crumbs::{Located, InfixCrumb};
+use ast::macros::DocCommentInfo;
+
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation.rs]: Fix imports

File: src/rust/ide/src/double_representation.rs [hunk 6]
```diff
@@ -28,3 +36,107 @@ pub mod test_utils;
 ///
 /// Link: https://github.com/enso-org/enso/blob/main/doc/syntax/encoding.md
 pub const INDENT : usize = 4;
+
+/// What kind of node or definition a line should be treated as.
+#[derive(Clone,Debug)]
+pub enum LineKind {
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation.rs:54]: This is in a wrong section: `Consts`

File: src/rust/ide/src/double_representation.rs [hunk 7]
```diff
@@ -28,3 +36,107 @@ pub mod test_utils;
...
+    /// Definition is a binding, which defines a new entity with arguments.
+    Definition {
+        /// Ast of the whole binding.
+        ast  : known::Infix,
+        /// Definition name. If this is an extension method, it might imply an implicit argument.
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation.rs]: Mention implicit self explicitly.

File: src/rust/ide/src/double_representation.rs [hunk 8]
```diff
@@ -28,3 +36,107 @@ pub mod test_utils;
...
+    DocumentationComment {
+        /// The comment representation.
+        comment : DocCommentInfo,
+    }
+}
+
+impl LineKind {
+
+    /// Tell how the given line (described by Ast) should be treated.
+    /// Returns `None` if line does not match any of the expected entities.
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation.rs]: It  cannot return None.

File: src/rust/ide/src/double_representation.rs [hunk 9]
```diff
@@ -28,3 +36,107 @@ pub mod test_utils;
...
+    pub fn discern(ast:&Ast, kind:ScopeKind) -> Self {
+        use LineKind::*;
+
+        // First of all, if line is not an infix assignment, it can be only node or nothing.
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation.rs]: update comment (no nothing anymore).

File: src/rust/ide/src/double_representation.rs [hunk 10]
```diff
@@ -28,3 +36,107 @@ pub mod test_utils;
...
+            if is_setter || is_node {
+                return ExpressionAssignment{ast:infix}
+            }
+        };
+
+        Definition {
+            args,name,ast:infix.clone_ref()
+        }
+    }
+}
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation.rs:143]: This is worth testing here (not necessarily now, but at least TODO)

File: src/rust/ide/src/double_representation/definition.rs [hunk 11]
```diff
@@ -7,9 +7,9 @@ use ast::crumbs::Crumbable;
 use ast::crumbs::InfixCrumb;
 use ast::crumbs::Located;
 use ast::known;
-use ast::prefix;
 use ast::opr;
 use parser::Parser;
+use crate::double_representation::LineKind;
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation/definition.rs]: Import ordering.

File: src/rust/ide/src/double_representation/definition.rs [hunk 12]
```diff
@@ -403,7 +367,10 @@ impl<'a> DefinitionIterator<'a> {
     /// Looks up direct child definition by given name.
     pub fn find_by_name(mut self, name:&DefinitionName) -> Result<ChildDefinition,CannotFindChild> {
         let err = || CannotFindChild(name.clone());
-        self.find(|child_def| &*child_def.item.name == name).ok_or_else(err)
+        self.find(|child_def| {
+            INFO!(&*child_def.item.name);
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation/definition.rs]: Debug log

File: src/rust/ide/src/double_representation/node.rs [hunk 13]
```diff
@@ -5,6 +5,11 @@ use crate::prelude::*;
 use ast::Ast;
 use ast::crumbs::Crumbable;
 use ast::known;
+use std::cmp::Ordering;
+use ast::macros::DocCommentInfo;
+use crate::double_representation::LineKind;
+use crate::double_representation::definition::ScopeKind;
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation/node.rs]: Imports

File: src/rust/ide/src/double_representation/node.rs [hunk 14]
```diff
@@ -19,23 +24,106 @@ pub type Id = ast::Id;
...
+    /// Main line is a line that contains the node's expression.
+    pub main_line          : usize,
+}
+
+impl PartialOrd for NodeIndex {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.main_line.partial_cmp(&other.main_line)
+    }
+}
+
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation/node.rs]: This implementation does not agree with derived PartialEq.

File: src/rust/ide/src/double_representation/node.rs [hunk 15]
```diff
@@ -19,23 +24,106 @@ pub type Id = ast::Id;
 #[fail(display="Node with ID {} was not found.", id)]
 pub struct IdNotFound {pub id:Id}
 
+/// Indices of lines belonging to a node.
+#[derive(Clone,Copy,Debug,PartialEq)]
+pub struct NodeIndex {
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation/node.rs]: NodeLocation is better.

File: src/rust/ide/src/double_representation/node.rs [hunk 16]
```diff
@@ -287,7 +466,7 @@ mod tests {
         let larg     = Ast::var("foo");
         let rarg     = Ast::var("bar").with_id(id);
         let line_ast = Ast::infix(larg,ASSIGNMENT,rarg);
-        let mut node = NodeInfo::from_line_ast(&line_ast).unwrap();
+        let mut node = NodeInfo::from_main_line_ast(&line_ast).unwrap();
 
         assert_eq!(node.repr(), "foo = bar");
         assert_eq!(node.id(),id);
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation/node.rs:498]: Test for iterating over nodes in line list.

File: src/rust/ide/src/double_representation/graph.rs [hunk 17]
```diff
@@ -61,8 +60,14 @@ impl GraphInfo {
         let body = ast.rarg.clone();
         if let Ok(body_block) = known::Block::try_new(body.clone()) {
             block_nodes(&body_block)
+        } else if let Some(main_line) = node::MainLine::from_ast(&body) {
+            // There's no way to attach a documentation comment to an inline node.
+            let documentation = None;
+            vec![ NodeInfo {documentation,main_line} ]
         } else {
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation/graph.rs]: Use the helper for that.

File: src/rust/ide/src/controller/graph.rs [hunk 18]
```diff
@@ -1271,83 +1291,83 @@ main =
...
+//             };
+//             graph.add_node(info_w_pattern).unwrap();
+//             let expected_program = r"
+// main =
+//     foo = 2
+//     print foo
+//     sum1 = a+b";
+//             model::module::test::expect_code(&*graph.module,expected_program);
+//         })
+//     }
```
Comment (farmaazon) [inline src/rust/ide/src/controller/graph.rs]: Restore commented tests.

File: src/rust/ide/view/graph-editor/src/component/node.rs [hunk 19]
```diff
@@ -69,6 +69,14 @@ const UNRESOLVED_SYMBOL_TYPE : &str = "Builtins.Main.Unresolved_Symbol";
 
 
 
+// ===============
+// === Comment ===
+// ===============
+
+pub type Comment = String;
```
Comment (farmaazon) [inline src/rust/ide/view/graph-editor/src/component/node.rs:83]: Missing docs + mention why string and not ImString.

File: src/rust/ide/lib/ast/impl/src/macros.rs (hunk 1 above)
Comment (mwu-tow) [reply src/rust/ide/lib/ast/impl/src/macros.rs]: Nope, without `r` it becomes a comment.

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 20]
```diff
@@ -630,7 +640,6 @@ pub struct BlockLine <T> {
 }
 
 
-
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/lib.rs:633]: restore line.

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 21]
```diff
@@ -1201,29 +1210,68 @@ impl<T> BlockLine<T> {
...
+    }
+
+    pub fn transpose_ref(&self) -> Option<BlockLine<&T>> {
+        self.as_ref().map(Option::as_ref).transpose()
+    }
+    
+    pub fn map_opt<U>(self, f:impl FnOnce(T) -> U) -> BlockLine<Option<U>> {
+        self.map(|elem| elem.map(f))
+    } 
+}
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/lib.rs]: Do you have warnings about missing documentation?

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 22]
```diff
@@ -1201,29 +1210,68 @@ impl<T> BlockLine<T> {
...
-        let off        = first_line.off;
-        lines.push(BlockLine{elem,off});
-        lines.extend(self.lines.iter().cloned());
-        lines
+    /// Iterate over non-empty lines, while keeping their absolute indices.
+    pub fn enumerate_non_empty_lines(&self) -> impl Iterator<Item=(usize,BlockLine<&T>)> + '_ {
+        self.iter_all_lines().enumerate().filter_map(|(index,line):(usize,BlockLine<Option<&T>>)| {
+            let non_empty_line = line.transpose()?;
+            Some((index, non_empty_line))
+        })
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/lib.rs:1278]: Why we cannot use non-associated enumerate_non_empty_lines here?

File: src/rust/ide/src/controller/graph.rs [hunk 23]
```diff
@@ -664,7 +657,10 @@ impl Handle {
                     false
                 }
             };
-            let range = NodeIndex::range(node_to_be_after.index,node_to_be_before.index);
+
+            // FIXME block lines are not necessarily in a block
```
Comment (farmaazon) [inline src/rust/ide/src/controller/graph.rs]: Fixme left.

**Review by farmaazon — APPROVED** [review]

**Pull request merged**